File: bindings/python/tests/fixtures/readme_quickstart.py
"""The quick start from ``bindings/python/README.md``, verbatim.

``test_readme_quickstart.py`` runs this file as its own process with
``SCP_KEY_PASSPHRASE`` and a scratch ``HOME`` set, which is what a reader does
from a shell. A change to either copy that the other does not mirror fails
review, so the README stops drifting from what runs.
"""

import asyncio

from scp_sdk import SCP, Capability, CustodyType, MemoryScope


async def main():
    # Every call routes through an SCP instance (ADR-048). Name a storage
    # backend: this constructor has no default.
    scp = SCP(storage={"type": "in_memory"})

    # Create a cryptographic identity (DID). Name a custody backend too —
    # `identity_create` has no default either (spec §17.17.1,
    # SCP-CAPSEL-8000). `CustodyType.FILE` encrypts $HOME/.scp/keys.bin under
    # SCP_KEY_PASSPHRASE (Argon2id + AES-256-GCM, spec §17.8).
    identity = await scp.identity_create(CustodyType.FILE)
    print(f"DID: {identity.did}")

    # Create an encrypted context. The ceiling bounds every capability any
    # member of this context can ever hold, so it must carry `context:close`
    # for the `context_close` call below to pass its capability check.
    ctx = await scp.context_create(
        identity.did,
        {
            "ceiling": [
                Capability.MESSAGES_READ.value,
                Capability.MESSAGES_WRITE.value,
                Capability.CONTEXT_CLOSE.value,
            ],
            "memory_scope": MemoryScope.EPHEMERAL.value,
            "ttl": 3600,
        },
    )

    # Send a message (MLS-encrypted, signed, provenance-tagged).
    await scp.context_send(ctx._raw_handle, identity.did, b"Hello from SCP")

    await scp.context_close(ctx._raw_handle, identity.did)
    await scp.shutdown(5.0)


asyncio.run(main())

File: bindings/typescript/tests/fixtures/readme-quickstart.ts
// The quick start from `bindings/typescript/README.md`, verbatim apart from the
// import path (the README names the published package; this file names the
// source it is published from).
//
// `readme-quickstart.test.ts` spawns this file with `SCP_KEY_PASSPHRASE` and a
// scratch `HOME` set, because the Rust side reads the NATIVE process
// environment and a `process.env` write inside a running Bun process does not
// reach it. A change to either copy that the other does not mirror fails
// review, so the README stops drifting from what runs.

import { SCP } from "../../src/scp";

// Every call routes through an SCP instance (ADR-048). Name a storage
// backend: this constructor has no default.
const scp = new SCP({ storage: { type: "in_memory" } });

// Create a cryptographic identity (DID). Name a custody backend too —
// `identityCreate` has no default either (spec §17.17.1, SCP-CAPSEL-8000).
// `"file"` encrypts $HOME/.scp/keys.bin under SCP_KEY_PASSPHRASE (Argon2id +
// AES-256-GCM, spec §17.8), so this process owns its keys with no injected
// provider. `"platform"` and `"software"` instead require a
// KeyCustodyProvider you wire through `identityCreateWithCustody`.
const identity = await scp.identityCreate("file");
console.log(`DID: ${identity.did}`);

// Create an encrypted context. The ceiling bounds every capability any member
// of this context can ever hold, so it must carry `context:close` for the
// `contextClose` call below to pass its capability check.
const ctx = await scp.contextCreate(
  identity,
  JSON.stringify({
    ceiling: ["messages:read", "messages:write", "context:close"],
    memoryScope: "ephemeral",
    governance: "single_admin",
    ttl: 3600,
  }),
);

// Send a message (MLS-encrypted, signed, provenance-tagged).
await scp.contextSend(
  ctx._rawHandle,
  identity.did,
  new TextEncoder().encode("Hello from SCP"),
  null,
);

await scp.contextClose(ctx._rawHandle, identity.did);
await scp.shutdown(5);
